src/LightningDB/LightningTransaction.cs

```diff
@@ -142,54 +156,140 @@ public byte[] Get(LightningDatabase db, byte[] key)
         }
 
         /// <summary>
         /// Check whether data exists in database.
         /// </summary>
-        /// <param name="db">Database.</param>
-        /// <param name="key">Key.</param>
+        /// <param name="db">The database to query.</param>
+        /// <param name="key">A span containing the key to look up.</param>
         /// <returns>True if key exists, false if not.</returns>
         public bool ContainsKey(LightningDatabase db, byte[] key)
+        {//argument validation delegated to next call
+            return ContainsKey(db, key.AsSpan());
+        }
+
+        /// <summary>
+        /// Check whether data exists in database.
+        /// </summary>
+        /// <param name="db">The database to query.</param>
+        /// <param name="key">A span containing the key to look up.</param>
+        /// <returns>True if key exists, false if not.</returns>
+        public bool ContainsKey(LightningDatabase db, ReadOnlySpan<byte> key)
         {
             if (db == null)
                 throw new ArgumentNullException(nameof(db));
 
+#warning this copies data when it doesn't need to...
             return TryGet(db, key, out _);
         }
 
+        /// <summary>
+        /// Put data into a database.
+        /// </summary>
+        /// <param name="db">The database to query.</param>
+        /// <param name="key">A span containing the key to look up.</param>
+        /// <param name="value">A byte array containing the value found in the database, if it exists.</param>
+        /// <param name="options">Operation options (optional).</param>
+        public void Put(LightningDatabase db, byte[] key, byte[] value, PutOptions options = PutOptions.None)
+        {//argument validation delegated to next call
+
+            Put(db, key.AsSpan(), value.AsSpan(), options);
+        }
+
+
         /// <summary>
         /// Put data into a database.
         /// </summary>
         /// <param name="db">Database.</param>
         /// <param name="key">Key byte array.</param>
         /// <param name="value">Value byte array.</param>
         /// <param name="options">Operation options (optional).</param>
-        public void Put(LightningDatabase db, byte[] key, byte[] value, PutOptions options = PutOptions.None)
+        public unsafe void Put(LightningDatabase db, ReadOnlySpan<byte> key, ReadOnlySpan<byte> value, PutOptions options = PutOptions.None)
         {
             if (db == null)
                 throw new ArgumentNullException(nameof(db));
 
-            var mdbKey = new MDBValue(new Span<byte>(key));
-            var mdbValue = new MDBValue(new Span<byte>(value));
-            mdb_put(_handle, db.Handle(), mdbKey, mdbValue, options);
+            fixed (byte* keyPtr = key)
+            fixed (byte* valuePtr = value)
+            {
+                var mdbKey = new MDBValue(key.Length, keyPtr);
+                var mdbValue = new MDBValue(value.Length, valuePtr);
+
+#warning surface return code?

```

Comment:
It probably doesn't hurt to add the return code, but I'm inclined to say if we do that we should also re-evaluate the exception behavior. For now, let's leave it out unless you have a strong argument for adding it.

Reply:
I'm going to make a separate issue for this